File: test/src/tc/ic/junit/src/org/iotivity/test/ic/tc/stc/ICMessageQueueTest.java
/******************************************************************
 *
 * Copyright 2017 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************/

package org.iotivity.test.ic.tc.stc;

import java.util.EnumSet;
import java.util.List;

import android.test.InstrumentationTestCase;
import android.content.Context;
import android.test.AndroidTestCase;
import android.util.Log;

import org.iotivity.base.OcPlatform;
import org.iotivity.base.OcConnectivityType;
import org.iotivity.base.OcException;
import org.iotivity.base.OcHeaderOption;
import org.iotivity.base.ErrorCode;
import org.iotivity.base.ModeType;
import org.iotivity.base.OcAccountManager;
import org.iotivity.base.OcPresenceHandle;
import org.iotivity.base.OcRepresentation;
import org.iotivity.base.PlatformConfig;
import org.iotivity.base.QualityOfService;
import org.iotivity.base.ResourceProperty;
import org.iotivity.base.ServiceType;
import org.iotivity.base.OcResource;
import org.iotivity.cloud.*;
import org.iotivity.configuration.*;
import org.iotivity.test.ic.tc.helper.ICHelper;
import org.iotivity.test.ic.tc.helper.ICResourceDirectoryCommonAdapter;
import org.iotivity.test.ic.tc.helper.OcAccountManagerAdapter;
import org.iotivity.test.ic.tc.helper.*;

import static org.iotivity.test.ic.tc.helper.ICHelperStaticUtil.*;

public class ICMessageQueueTest extends InstrumentationTestCase
        implements IConfiguration {

    private static final String    ALL_INTERFACE_TYPE = "0.0.0.0";
    private ICMessageAdapter       mIcMessageAdapter;
    QualityOfService               mQualityOfService;
    private Context                mContext;
    OcRepresentation               mRep               = new OcRepresentation();
    ICMessageQueue                 mMessageQueueHelper;
    private OcAccountManagerHelper mCloudHelper;
    private OcAccountManager       mAccountManager;

    protected void setUp() throws Exception {
        super.setUp();
        mMessageQueueHelper = new ICMessageQueue();
        mQualityOfService = QualityOfService.LOW;
        mIcMessageAdapter = new ICMessageAdapter();

        mContext = getInstrumentation().getContext();
        mCloudHelper = new OcAccountManagerHelper();

        PlatformConfig cfg = new PlatformConfig(
                getInstrumentation().getTargetContext(), ServiceType.IN_PROC,
                ModeType.CLIENT_SERVER, ALL_INTERFACE_TYPE, // bind to all
                                                            // available
                // interfaces
                0, QualityOfService.LOW);
        OcPlatform.Configure(cfg);

        mAccountManager = OcAccountManagerHelper
                .getAccountMangerInstance(TLS.DISABLED);
        OcAccountManagerHelper.init(mContext);

        MY_MAP.put("rt", "light");

        // signUp and SignIn for all TC
        assertTrue(mCloudHelper.singUp(mAccountManager, DEFAULT_AUTH_PROVIDER,
                OcAccountManagerHelper.authCode, mCloudHelper));

        assertTrue(mCloudHelper.singIn(mAccountManager,
                OcAccountManagerHelper.s_CloudUid,
                OcAccountManagerHelper.s_CloudAccesstoken, mCloudHelper));

        mMessageQueueHelper.getMQBroker();
    }

    protected void tearDown() throws Exception {
        super.tearDown();
    }

    /**
     * @since 2017-01-26
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'createMQTopic' API With valid scenario
     * @target createMQTopic(const OCRepresentation& rep, const std::string&
     *         topicUri, const QueryParamsMap& queryParametersMap,
     *         MQTopicCallback attributeHandler, QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn
     * @procedure Call createMQTopic API with valid parameter
     * @post_condition SignOut
     * @expected createMQTopic API works properly without any exception..
     */
    public void testCreateMQTopic_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-01-26
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'createMQTopic' API With valid scenario
     * @target createMQTopic(const OCRepresentation& rep, const std::string&
     *         topicUri, const QueryParamsMap& queryParametersMap,
     *         MQTopicCallback attributeHandler, QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn
     * @procedure Call createMQTopic API with valid parameter
     * @post_condition SignOut
     * @expected createMQTopic API works properly without any exception..
     */
    public void testCreateMQTopicByType_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("currentTopicResource is null",
                    ICMessageAdapter.scurrentTopicResource);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-01-26
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'discoveryMQTopics' API With valid scenario
     * @target OCStackResult discoveryMQTopics(const QueryParamsMap&
     *         queryParametersMap, MQTopicCallback attributeHandler,
     *         QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn 4. Create Topic
     * @procedure Call discoveryMQTopics API with valid parameter
     * @post_condition SignOut
     * @expected discoveryMQTopics API works properly without any exception..
     */
    public void testDiscoveryMQTopics_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);

            sMQbrokerResource.discoveryMQTopics(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-01-26
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'publishMQTopic' API With valid scenario
     * @target OCStackResult publishMQTopic(const QueryParamsMap&
     *         queryParametersMap, MQTopicCallback
     *         attributeHandler,QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn
     * @procedure 1. Perform createMQTopic() API 2. Perform discoveryMQTopics()
     *            API 3. Call publishMQTopic API with valid parameter 4. Check
     *            all APIs work properly
     * @post_condition SignOut
     * @expected publishMQTopic API works properly without any exception..
     */
    public void testdiscoveryMQTopicsByType_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("currentTopicResource is null",
                    ICMessageAdapter.scurrentTopicResource);

            sMQbrokerResource.discoveryMQTopics(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("resourceUri is null", ICMessageAdapter.sResourceUri);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-01-26
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'publishMQTopic' API With valid scenario
     * @target OCStackResult publishMQTopic(const QueryParamsMap&
     *         queryParametersMap, MQTopicCallback
     *         attributeHandler, QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn
     * @procedure 1. Perform createMQTopic() API 2. Perform discoveryMQTopics()
     *            API 3. Call publishMQTopic API with valid parameter 4. Check
     *            all APIs work properly
     * @post_condition SignOut
     * @expected publishMQTopic API works properly without any exception..
     */
    public void testPublishMQTopic_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("currentTopicResource is null",
                    ICMessageAdapter.scurrentTopicResource);

            sMQbrokerResource.discoveryMQTopics(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("resourceUri is null", ICMessageAdapter.sResourceUri);

            mRep = mMessageQueueHelper.generateRep();
            assertNotNull("m_rep is null", mRep);

            sMQbrokerResource.publishMQTopic(mRep, MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnPostCompleted is failed",
                    ICMessageAdapter.sOnPostCompleted);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-01-26
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'publishMQTopic' API With valid scenario
     * @target OCStackResult publishMQTopic(const QueryParamsMap&
     *         queryParametersMap, MQTopicCallback attributeHandler,
     *         QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn
     * @procedure 1. Perform createMQTopic() API 2. Perform discoveryMQTopics()
     *            API 3. Call requestMQPublish API with valid parameter 4. Check
     *            all APIs work properly
     * @post_condition SignOut
     * @expected publishMQTopic API works properly without any exception..
     */
    public void testRequestpublishMQTopic_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("currentTopicResource is null",
                    ICMessageAdapter.scurrentTopicResource);

            sMQbrokerResource.discoveryMQTopics(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("resourceUri is null", ICMessageAdapter.sResourceUri);

            mRep = mMessageQueueHelper.generateRep();
            assertNotNull("m_rep is null", mRep);

            sMQbrokerResource.requestMQPublish(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnPostCompleted is failed",
                    ICMessageAdapter.sOnPostCompleted);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-02-03
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'subscribeMQTopic' API With valid scenario
     * @target OCStackResult subscribeMQTopic(const QueryParamsMap&
     *         queryParametersMap,MQTopicCallback attributeHandler,
     *         QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn 4. discover Topic
     * @procedure 1. Perform createMQTopic() API 2. Perform discoveryMQTopics()
     *            API 3. Call subscribeMQTopic API with valid parameter 4. Check
     *            all APIs work properly
     * @post_condition SignOut
     * @expected subscribeMQTopic API works properly without any exception..
     */
    public void testSubscribeMQTopic_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("currentTopicResource is null",
                    ICMessageAdapter.scurrentTopicResource);

            sMQbrokerResource.discoveryMQTopics(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("resourceUri is null", ICMessageAdapter.sResourceUri);

            mRep = mMessageQueueHelper.generateRep();
            assertNotNull("m_rep is null", mRep);

            ICMessageAdapter.scurrentTopicResource.subscribeMQTopic(MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnPostCompleted is failed",
                    ICMessageAdapter.sOnObserveCompleted);
            assertNotNull("resourceUri is null",
                    ICMessageAdapter.sCurrentResourceRep);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

    /**
     * @since 2017-02-03
     * @see public void signUp(String authProvider, String authCode,
     *      OnPostListener onPostListener)
     * @see public void signIn(String userUuid, String accessToken,
     *      OnPostListener onPostListener)
     * @see public void signOut(String accessToken, OnPostListener
     *      onPostListener)
     * @objective Test 'unsubscribeMQTopic' API With valid scenario
     * @target OCStackResult unsubscribeMQTopic(QualityOfService qos)
     * @test_data 1. OCRepresentation 2. topicUri 3. queryParametersMap 4.
     *            MQTopicCallback 5. QualityOfService
     * @pre_condition 1. constructAccountManagerObject(host, connectivity_type)
     *                API 2. SignUp 3. SignIn 4. discover Topic
     * @procedure 1. Perform createMQTopic() API 2. Perform discoveryMQTopics()
     *            API 3. Call subscribeMQTopic() and unsubscribeMQTopic() APIs
     *            with valid parameter 4. Check all APIs work properly
     * @post_condition SignOut
     * @expected unsubscribeMQTopic API works properly without any exception..
     */
    public void testUnsubscribeMQTopic_SRC_P() {
        try {
            assertNotNull("MQbrokerResource is null", sMQbrokerResource);
            sMQbrokerResource.createMQTopic(mRep, MQ_DEFAULT_TOPIC_URI, MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("OnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("currentTopicResource is null",
                    ICMessageAdapter.scurrentTopicResource);

            sMQbrokerResource.discoveryMQTopics(MY_MAP, mIcMessageAdapter,
                    IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnTopicResourceCreated is failed",
                    ICMessageAdapter.sOnTopicResourceCreated);
            assertNotNull("resourceUri is null", ICMessageAdapter.sResourceUri);

            ICMessageAdapter.scurrentTopicResource.subscribeMQTopic(MY_MAP,
                    mIcMessageAdapter, IC_OC_QUALITY_OF_SERVICE);
            assertTrue("mOnPostCompleted is failed",
                    ICMessageAdapter.sOnObserveCompleted);
            assertNotNull("resourceUri is null",
                    ICMessageAdapter.sCurrentResourceRep);

            ICMessageAdapter.scurrentTopicResource
                    .unsubscribeMQTopic(IC_OC_QUALITY_OF_SERVICE);
        } catch (OcException ex) {
            fail("Exception occurred: " + ex.getLocalizedMessage());
        }
    }

}

File: test/src/tc/ic/gtest/src/btc/ICMessageQueueTest.cpp
/******************************************************************
*
* Copyright 2017 Samsung Electronics All Rights Reserved.
*
*
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
******************************************************************/

#include "ICHelper.h"

using namespace OC;

class ICMessageQueueTest_btc: public ::testing::Test{

public:
    std::vector< std::string > m_types = { "oic.ps" };
    std::vector< std::string > m_iFaces = { DEFAULT_INTERFACE };

    OC::OCResource::Ptr m_OCResource ;

    QueryParamsMap m_queryParametersMap;
    QualityOfService m_qos;
    OCRepresentation m_rep;
    ObserveType m_observe;
    QueryParamsMap m_query;

    static void foundResource(const int ecode, const string &originUri,
            shared_ptr< OC::OCResource > topic)
    {

    }

    static void observeHandler(const HeaderOptions&, const OCRepresentation&, const int, const int)
    {

    }

    static void attributeHandler(const HeaderOptions&, const OCRepresentation&, const int)
    {

    }

protected:
    std::string m_errorMsg;

    virtual void SetUp()
    {
        m_OCResource = OCPlatform::constructResourceObject(HOST, URI, CT_DEFAULT, true, m_types, m_iFaces);
        m_query = {};
        m_qos = QualityOfService::LowQos;
    }

    virtual void TearDown()
    {
    }
};

/**
 * @since 2016-08-25
 * @see None
 * @objective Test createMQTopic function with valid OCRepresentation, topicURI QueryMap, createCallback and QOS
 * @target createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback)
 * @test_data OCRepresentation, topic uri, query params, createCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback) API
 * @post_condition None
 * @expected createMQTopic should return OC_STACK_OK
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, CreateMQTopic_SRC_P)
{
    EXPECT_EQ(OC_STACK_OK, m_OCResource->createMQTopic(m_rep, TOPIC_URI, m_queryParametersMap, &foundResource, m_qos))<<"CreateMQTopic is not creating Topic into MQ Broker";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test createMQTopic function with NULL callback
 * @target createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback)
 * @test_data OCRepresentation, topic uri, query params, NULL createCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback) API
 * @post_condition None
 * @expected createMQTopic will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, CreateMQTopicNULLCallback_NV_N)
{
    EXPECT_ANY_THROW(m_OCResource->createMQTopic(m_rep, TOPIC_URI, m_queryParametersMap, nullptr, m_qos))<<"CreateMQTopic is creating Topic into MQ Broker with NULL callback";
}
#endif

/**
 * @since 2016-08-26
 * @see None
 * @objective Test createMQTopic function with empt topicURI
 * @target createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback)
 * @test_data OCRepresentation, empty topic uri, query params, createCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback) API
 * @post_condition None
 * @expected createMQTopic will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, CreateMQTopic_ESV_N)
{
    EXPECT_ANY_THROW(m_OCResource->createMQTopic(m_rep, EMPTY_STRING, m_queryParametersMap, &foundResource, m_qos))<<"CreateMQTopic is creating Topic into MQ Broker with empty string as topic uri";
}
#endif

/**
 * @since 2016-08-26
 * @see None
 * @objective Test createMQTopic function with invalid topicURI
 * @target createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback)
 * @test_data OCRepresentation, invalid topic uri, query params, createCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform createMQTopic(const OCRepresentation& rep, const std::string& topicUri, const QueryParamsMap& queryParametersMap, &createCallback) API
 * @post_condition None
 * @expected createMQTopic will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, CreateMQTopic_USV_N)
{
    EXPECT_ANY_THROW(m_OCResource->createMQTopic(m_rep, INVALID, m_queryParametersMap, &foundResource, m_qos))<<"CreateMQTopic is creating Topic into MQ Broker with invalid string as topic uri";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test 'discoveryMQTopics' function with valid QueryParameterMap, createCallback and QOS
 * @target discoveryMQTopics(const QueryParamsMap& queryParametersMap, &findCallback)
 * @test_data queryParameterMap, findCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform discoveryMQTopics(const QueryParamsMap& queryParametersMap, &findCallback) API
 * @post_condition None
 * @expected discoveryMQTopics that provide OC_STACK_OK
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, DiscoveryMQTopics_SRC_P)
{
    EXPECT_EQ(OC_STACK_OK,m_OCResource->discoveryMQTopics(m_queryParametersMap, &foundResource, m_qos))<<"DiscoveryMQTopics is not to discovery Topics from MQ Broker.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test 'discoveryMQTopics' function with NULL Callback
 * @target discoveryMQTopics(const QueryParamsMap& queryParametersMap, &findCallback)
 * @test_data queryParameterMap, NULL findCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform discoveryMQTopics(const QueryParamsMap& queryParametersMap, &findCallback) API
 * @post_condition None
 * @expected discoveryMQTopics will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, DiscoveryMQTopicsCallback_NV_N)
{
    EXPECT_ANY_THROW(m_OCResource->discoveryMQTopics(m_queryParametersMap, nullptr, m_qos))<<"DiscoveryMQTopics is to discovery Topics from MQ Broker with NULL callback.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test requestMQPublish API with valid QueryParameterMap, publishCallback and QOS
 * @target requestMQPublish(const QueryParamsMap& queryParametersMap, &publishCallback)
 * @test_data queryParameterMap, publishCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform requestMQPublish(const QueryParamsMap& queryParametersMap, &publishCallback) API
 * @post_condition None
 * @expected requestMQPublish should return OC_STACK_OK
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, RequestMQPublish_SRC_P)
{
    EXPECT_EQ(OC_STACK_OK, m_OCResource->requestMQPublish(m_queryParametersMap, &attributeHandler, m_qos))<<"RequestMQPublish is not to request publish to MQ publisher.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test requestMQPublish API with NULL Callback
 * @target requestMQPublish(const QueryParamsMap& queryParametersMap, &publishCallback)
 * @test_data queryParameterMap, NULL publishCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform requestMQPublish(const QueryParamsMap& queryParametersMap, &publishCallback) API
 * @post_condition None
 * @expected requestMQPublish will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, RequestMQPublishNULLCallback_NV_N)
{
    EXPECT_ANY_THROW(m_OCResource->requestMQPublish(m_queryParametersMap, nullptr, m_qos))<<"RequestMQPublish is to request publish to MQ publisher with NULL callback.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test subscribeMQTopic API with valid observer, QueryParameterMap, observeCallback and QOS
 * @target subscribeMQTopic(ObserveType observeType, const QueryParamsMap& queryParametersMap, &observeCallback)
 * @test_data observeType, queryParameterMap, observeCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform subscribeMQTopic(ObserveType observeType, const QueryParamsMap& queryParametersMap, &observeCallback) API
 * @post_condition None
 * @expected subscribeMQTopic should return OC_STACK_OK
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, SubscribeMQTopic_SRC_P)
{
    EXPECT_EQ(OC_STACK_OK, m_OCResource->subscribeMQTopic(m_observe, m_queryParametersMap, &observeHandler, m_qos))<<"SubscribeMQTopic is not to subscribe Topic to MQ Broker.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test subscribeMQTopic API with NULL Callback
 * @target subscribeMQTopic(ObserveType observeType, const QueryParamsMap& queryParametersMap, &observeCallback)
 * @test_data observeType, queryParameterMap, NULL observeCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform subscribeMQTopic(ObserveType observeType, const QueryParamsMap& queryParametersMap, &observeCallback) API
 * @post_condition None
 * @expected subscribeMQTopic will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, SubscribeMQTopicNULLCallback_NV_N)
{
    EXPECT_ANY_THROW(m_OCResource->subscribeMQTopic(m_observe, m_queryParametersMap, nullptr, m_qos))<<"SubscribeMQTopic is to subscribe Topic to MQ Broker with NULL callback.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test unsubscribeMQTopic API with valid QOS
 * @target unsubscribeMQTopic()
 * @test_data qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform unsubscribeMQTopic() API
 * @post_condition None
 * @expected unsubscribeMQTopic should return OC_STACK_OK
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, UnsubscribeMQTopic_SRC_P)
{
    EXPECT_EQ(OC_STACK_OK,m_OCResource->subscribeMQTopic(m_observe, m_queryParametersMap, observeHandler, m_qos))<<"SubscribeMQTopic is NOT to subscribe Topic to MQ Broker with NULL callback.";
    EXPECT_EQ(OC_STACK_OK, m_OCResource->unsubscribeMQTopic(m_qos))<<"UnsubscribeMQTopic is not to unsubscribe Topic to MQ Broker.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test publishMQTopic API with valid OCRepresentation, QueryParameterMap, publishCallback and QOS
 * @target publishMQTopic(const OCRepresentation& rep, const QueryParamsMap& queryParametersMap, &publishCallback)
 * @test_data OCRepresentation, queryParameterMap, publishCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform publishMQTopic(const OCRepresentation& rep, const QueryParamsMap& queryParametersMap, &publishCallback) API
 * @post_condition None
 * @expected publishMQTopic should return OC_STACK_OK
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, PublishMQTopic_SRC_P)
{
    EXPECT_EQ(OC_STACK_OK, m_OCResource->publishMQTopic(m_rep, m_queryParametersMap, &attributeHandler, m_qos))<<"PublishMQTopic is not to publish Topic information into MQ Broker.";
}
#endif

/**
 * @since 2016-08-25
 * @see None
 * @objective Test publishMQTopic API with NULL Callback
 * @target publishMQTopic(const OCRepresentation& rep, const QueryParamsMap& queryParametersMap, &publishCallback)
 * @test_data OCRepresentation, queryParameterMap, NULL publishCallback and qos
 * @pre_condition constructResourceObject(host, uri, connectivity type, true, types, iFaces) API
 * @procedure Perform publishMQTopic(const OCRepresentation& rep, const QueryParamsMap& queryParametersMap, &publishCallback) API
 * @post_condition None
 * @expected publishMQTopic will throw Exception.
 **/
#if defined(__LINUX__) || defined(__TIZEN__)
TEST_F(ICMessageQueueTest_btc, PublishMQTopicNULLCall_NV_N)
{
    EXPECT_ANY_THROW(m_OCResource->publishMQTopic(m_rep, m_queryParametersMap, nullptr, m_qos))<<"PublishMQTopic is to publish Topic information into MQ Broker with NULL callback.";
}
#endif
